wallet/src/main/java/com/android/identity_credential/wallet/PresentationActivity.kt

```diff
@@ -217,159 +146,181 @@ class PresentationActivity : FragmentActivity() {
+                            ContextCompat.getMainExecutor(applicationContext),
+                            eDeviceKey!!
+                        )
+                        .useForwardEngagement(transport!!, deviceEngagement!!, handover!!)
+                        .build()
 
-                        State.CONNECTED -> {
-                            // on a new connected client, create a new DeviceRetrievalHelper and TransferHelper
-                            newDeviceRetrievalHelper().run {
-                                newTransferHelper(this)
-                            }
-                            stateDisplay.value = "Connected"
-                            Logger.i(TAG, "State: Connected")
-                        }
+                }
 
-                        State.REQUEST_AVAILABLE -> {
-                            stateDisplay.value = "Request Available"
-                            Logger.i(TAG, "State: Request Available")
-                            // start processing request and use processed request data to show consent prompt
-                            transferHelper?.startProcessingRequest(getDeviceRequest())
-                                ?.let { requestData ->
-                                    // update UI state object 'consentPromptData' so we can show ConsentPrompt
-                                    consentPromptData.value = ConsentPromptEntryFieldData(
-                                        credentialId = requestData.document.name,
-                                        documentName = requestData.document.documentConfiguration.displayName,
-                                        credentialData = requestData.document.documentConfiguration.mdocConfiguration!!.staticData,
-                                        documentRequest = requestData.documentRequest,
-                                        docType = requestData.docType,
-                                        verifier = requestData.trustPoint,
+                State.REQUEST_AVAILABLE -> {
+                    Logger.i(TAG, "State: Request Available")
+                    /**
+                     * Device request bytes have been transmitted via device retrieval helper listener,
+                     * Start the Presentment Flow where a Presentment is shown for every
+                     * [DocumentRequest] that has a suitable [Document].
+                     */
+                    lifecycleScope.launch {
+
+                        try {
+                            val deviceRequest = DeviceRequestParser(
+                                deviceRequestByteArray!!,
+                                deviceRetrievalHelper!!.sessionTranscript
+                            ).parse()
+
+                            deviceRequest.docRequests.forEach { docRequest ->
+                                // find a suitable Document for the docRequest
+                                val document = findSuitableDocumentForRequest(docRequest)
```

Comment:
I think it would be clearer if `findSuitableDocumentForRequest()` was called `findMdocCredentialForRequest()` and returned a `MdocCredential?` ... then I think you can just do
```
val mdocCredential = findMdocCredentialForRequest() ?: return@forEach
```
which I think is quite readable.

In the current behavior `findSuitableDocument()` throws if it can't find a document and this fails the entire transaction... which is not what we want, for example, suppose an RP asks for two docs (for example, mDL and COVID cert) ... you still want that transaction to succeed even if we can return one of the documents.

Reply:
Great idea for `findMdocCredentialForRequest` 👍  and you're absolutely right about the exception of `findSuitableDocument`, it should iterate through every `docRequest` even if the previous one had no matching MdocCredentials